File: check-connections.js
/**
 * Frontend Connection Verification Script
 * Tests API connectivity and external services from browser console
 * 
 * Usage: Include this file in your HTML or run in browser console
 * or use: node check-connections.js
 */

class ConnectionChecker {
    constructor() {
        this.results = {
            database: {},
            google: {},
            api: {},
            services: {}
        };
    }

    /**
     * Test API Backend Connection
     */
    async testBackendConnection() {
        console.log("=" .repeat(80));
        console.log("API BACKEND CONNECTION TEST");
        console.log("=" .repeat(80));
        
        const backendUrl = 'http://localhost:8000/api';
        
        try {
            const response = await fetch(`${backendUrl}/`, {
                method: 'GET',
                headers: {
                    'Accept': 'application/json',
                }
            });
            
            if (response.ok) {
                console.log("✅ Backend Server: CONNECTED");
                console.log(`Status: HTTP ${response.status}`);
                this.results.api.backend = {
                    status: '✅ Connected',
                    url: backendUrl,
                    statusCode: response.status
                };
            } else {
                console.log(`⚠️ Backend Server: HTTP ${response.status}`);
                this.results.api.backend = {
                    status: '⚠️ Error',
                    statusCode: response.status
                };
            }
        } catch (error) {
            console.log("❌ Backend Server: NOT CONNECTED");
            console.log(`Error: ${error.message}`);
            this.results.api.backend = {
                status: '❌ Failed',
                error: error.message
            };
        }
    }

    /**
     * Test API Endpoints
     */
    async testAPIEndpoints() {
        console.log("\n" + "=" .repeat(80));
        console.log("API ENDPOINTS TEST");
        console.log("=" .repeat(80));
        
        const endpoints = [
            { url: 'http://localhost:8000/api/auth/token/', method: 'POST', name: 'Authentication Token' },
            { url: 'http://localhost:8000/api/users/', method: 'GET', name: 'User Profile' },
            { url: 'http://localhost:8000/api/farms/', method: 'GET', name: 'Farms List' },
            { url: 'http://localhost:8000/api/disease-detection/', method: 'GET', name: 'Disease Detection' },
            { url: 'http://localhost:8000/api/weather/', method: 'GET', name: 'Weather Data' },
            { url: 'http://localhost:8000/api/alerts/', method: 'GET', name: 'Alerts' },
        ];
        
        this.results.api.endpoints = {};
        
        for (const endpoint of endpoints) {
            try {
                const response = await fetch(endpoint.url, {
                    method: endpoint.method,
                    headers: {
                        'Accept': 'application/json',
                        'Content-Type': 'application/json',
                    }
                });
                
                const status = response.ok ? '✅' : '⚠️';
                console.log(`${status} ${endpoint.name}: HTTP ${response.status}`);
                
                this.results.api.endpoints[endpoint.name] = {
                    status: status,
                    statusCode: response.status
                };
            } catch (error) {
                console.log(`❌ ${endpoint.name}: ${error.message}`);
                this.results.api.endpoints[endpoint.name] = {
                    status: '❌',
                    error: error.message
                };
            }
        }
    }

    /**
     * Test Google Services Reachability
     */
    async testGoogleServices() {
        console.log("\n" + "=" .repeat(80));
        console.log("GOOGLE SERVICES CONNECTIVITY TEST");
        console.log("=" .repeat(80));
        
        const googleServices = [
            { url: 'https://maps.google.com', name: 'Google Maps' },
            { url: 'https://cloud.google.com', name: 'Google Cloud' },
            { url: 'https://firebase.google.com', name: 'Firebase' },
            { url: 'https://accounts.google.com', name: 'Google OAuth' },
            { url: 'https://www.googleapis.com', name: 'Google APIs' },
        ];
        
        this.results.google.services = {};
        
        for (const service of googleServices) {
            try {
                // Using no-cors mode for external services
                const response = await fetch(service.url, {
                    method: 'HEAD',
                    mode: 'no-cors'
                });
                
                console.log(`✅ ${service.name}: REACHABLE`);
                this.results.google.services[service.name] = {
                    status: '✅ Reachable'
                };
            } catch (error) {
                console.log(`❌ ${service.name}: UNREACHABLE - ${error.message}`);
                this.results.google.services[service.name] = {
                    status: '❌ Unreachable',
                    error: error.message
                };
            }
        }
    }

    /**
     * Test External APIs
     */
    async testExternalAPIs() {
        console.log("\n" + "=" .repeat(80));
        console.log("EXTERNAL SERVICES TEST");
        console.log("=" .repeat(80));
        
        const services = [
            { url: 'https://api.openweathermap.org/data/2.5/weather', name: 'OpenWeatherMap API' },
            { url: 'https://api.weatherapi.com/v1/current.json', name: 'WeatherAPI' },
            { url: 'https://nominatim.openstreetmap.org/reverse', name: 'OpenStreetMap' },
            { url: 'https://ipapi.co/json/', name: 'IP Geolocation' },
        ];
        
        this.results.services.external = {};
        
        for (const service of services) {
            try {
                const response = await fetch(service.url, {
                    method: 'GET',
                    mode: 'cors'
                });
                
                const status = response.ok ? '✅ Accessible' : `⚠️ ${response.status}`;
                console.log(`${status} - ${service.name}`);
                
                this.results.services.external[service.name] = {
                    status: status,
                    statusCode: response.status
                };
            } catch (error) {
                console.log(`❌ ${service.name}: ${error.message}`);
                this.results.services.external[service.name] = {
                    status: '❌ Error',
                    error: error.message
                };
            }
        }
    }

    /**
     * Test LocalStorage and Session
     */
    testLocalStorage() {
        console.log("\n" + "=" .repeat(80));
        console.log("LOCAL STORAGE & SESSION TEST");
        console.log("=" .repeat(80));
        
        // Check LocalStorage
        try {
            const testKey = 'connection_test_' + Date.now();
            localStorage.setItem(testKey, 'test');
            const value = localStorage.getItem(testKey);
            
            if (value === 'test') {
                console.log("✅ LocalStorage: WORKING");
                localStorage.removeItem(testKey);
                this.results.services.localStorage = '✅ Working';
            } else {
                console.log("❌ LocalStorage: FAILED");
                this.results.services.localStorage = '❌ Failed';
            }
        } catch (error) {
            console.log(`❌ LocalStorage: ${error.message}`);
            this.results.services.localStorage = '❌ Error';
        }
        
        // Check SessionStorage
        try {
            const testKey = 'session_test_' + Date.now();
            sessionStorage.setItem(testKey, 'test');
            const value = sessionStorage.getItem(testKey);
            
            if (value === 'test') {
                console.log("✅ SessionStorage: WORKING");
                sessionStorage.removeItem(testKey);
                this.results.services.sessionStorage = '✅ Working';
            } else {
                console.log("❌ SessionStorage: FAILED");
                this.results.services.sessionStorage = '❌ Failed';
            }
        } catch (error) {
            console.log(`❌ SessionStorage: ${error.message}`);
            this.results.services.sessionStorage = '❌ Error';
        }
    }

    /**
     * Test Authentication Status
     */
    testAuthenticationStatus() {
        console.log("\n" + "=" .repeat(80));
        console.log("AUTHENTICATION STATUS");
        console.log("=" .repeat(80));
        
        const accessToken = localStorage.getItem('access_token');
        const refreshToken = localStorage.getItem('refresh_token');
        const userEmail = localStorage.getItem('user_email');
        
        if (accessToken) {
            console.log("✅ Access Token: PRESENT");
            console.log(`Token Length: ${accessToken.length} characters`);
            this.results.services.accessToken = '✅ Present';
        } else {
            console.log("⚠️ Access Token: NOT FOUND (User not logged in)");
            this.results.services.accessToken = '⚠️ Not Found';
        }
        
        if (refreshToken) {
            console.log("✅ Refresh Token: PRESENT");
            this.results.services.refreshToken = '✅ Present';
        } else {
            console.log("⚠️ Refresh Token: NOT FOUND");
            this.results.services.refreshToken = '⚠️ Not Found';
        }
        
        if (userEmail) {
            console.log(`✅ User Email: ${userEmail}`);
            this.results.services.userEmail = userEmail;
        } else {
            console.log("⚠️ User Email: NOT STORED");
        }
    }

    /**
     * Generate Summary Report
     */
    generateSummary() {
        console.log("\n" + "=" .repeat(80));
        console.log("SUMMARY REPORT");
        console.log("=" .repeat(80));
        
        console.log("\n📊 CONNECTION STATUS OVERVIEW:");
        console.log(`Backend API: ${this.results.api.backend?.status || 'Unknown'}`);
        console.log(`Local Storage: ${this.results.services.localStorage || 'Unknown'}`);
        console.log(`Authentication: ${this.results.services.accessToken || 'Not logged in'}`);
        
        console.log("\n📋 FULL RESULTS OBJECT:");
        console.log(JSON.stringify(this.results, null, 2));
        
        console.log("\n" + "=" .repeat(80));
        console.log("✅ CONNECTION VERIFICATION COMPLETE");
        console.log("=" .repeat(80));
    }

    /**
     * Run All Tests
     */
    async runAll() {
        console.clear();
        console.log("🔍 Starting Connection Verification...\n");
        
        await this.testBackendConnection();
        await this.testAPIEndpoints();
        await this.testGoogleServices();
        await this.testExternalAPIs();
        this.testLocalStorage();
        this.testAuthenticationStatus();
        this.generateSummary();
        
        return this.results;
    }
}

// Auto-run if in browser environment
if (typeof window !== 'undefined') {
    window.connectionChecker = new ConnectionChecker();
    console.log("ConnectionChecker available as: window.connectionChecker");
    console.log("Run: connectionChecker.runAll() to test all connections");
}

// Export for Node.js
if (typeof module !== 'undefined' && module.exports) {
    module.exports = ConnectionChecker;
}

File: check_connections.py
#!/usr/bin/env python
"""
Database and Google Connection Verification Script
Tests connectivity to Django database and Google services
"""

import os
import sys
import django
import json
from pathlib import Path

# Setup Django
sys.path.insert(0, str(Path(__file__).parent / 'backend'))
os.environ.setdefault('DJANGO_SETTINGS_MODULE', 'cropguard_backend.settings')
django.setup()

from django.db import connections
from django.db.utils import OperationalError
from django.contrib.auth.models import User
import requests

print("=" * 80)
print("CONNECTION VERIFICATION REPORT")
print("=" * 80)
print()

# ============================================
# 1. DATABASE CONNECTION TEST
# ============================================
print("📊 DATABASE CONNECTION TEST")
print("-" * 80)

db_results = {
    'status': 'unknown',
    'database': 'SQLite',
    'details': []
}

try:
    connection = connections['default']
    cursor = connection.cursor()
    cursor.execute("SELECT 1")
    result = cursor.fetchone()
    
    if result:
        db_results['status'] = '✅ CONNECTED'
        db_results['details'].append(f"Connection type: {connection.settings_dict['ENGINE']}")
        db_results['details'].append(f"Database path: {connection.settings_dict['NAME']}")
        
        # Check if database file exists
        db_path = connection.settings_dict['NAME']
        if os.path.exists(db_path):
            db_size = os.path.getsize(db_path)
            db_results['details'].append(f"Database file exists: Yes")
            db_results['details'].append(f"Database size: {db_size} bytes")
        
        # Count tables and records
        cursor.execute("SELECT name FROM sqlite_master WHERE type='table';")
        tables = cursor.fetchall()
        db_results['details'].append(f"Total tables: {len(tables)}")
        
        # Get table details
        table_info = {}
        for table in tables:
            table_name = table[0]
            cursor.execute(f"SELECT COUNT(*) FROM {table_name};")
            count = cursor.fetchone()[0]
            table_info[table_name] = count
        
        db_results['details'].append("Table records:")
        for table_name, count in table_info.items():
            db_results['details'].append(f"  - {table_name}: {count} records")
    else:
        db_results['status'] = '❌ FAILED'
        db_results['details'].append("Database query returned no result")
        
except OperationalError as e:
    db_results['status'] = '❌ FAILED'
    db_results['details'].append(f"OperationalError: {str(e)}")
except Exception as e:
    db_results['status'] = '❌ FAILED'
    db_results['details'].append(f"Error: {str(e)}")

for detail in db_results['details']:
    print(detail)
print()

# ============================================
# 2. GOOGLE SERVICES CONNECTION TEST
# ============================================
print("🔍 GOOGLE SERVICES CONNECTION TEST")
print("-" * 80)

google_results = {
    'Google Maps API': {'status': 'NOT CONFIGURED'},
    'Google Cloud Services': {'status': 'NOT CONFIGURED'},
    'Google OAuth': {'status': 'NOT CONFIGURED'},
    'Firebase': {'status': 'NOT CONFIGURED'},
}

# Test Google Maps API (geolocation)
print("Testing Google Maps API connectivity...")
try:
    # Using free Google Maps static API endpoint (doesn't require key for basic test)
    response = requests.head('https://maps.google.com', timeout=5)
    if response.status_code < 400:
        google_results['Google Maps API']['status'] = '✅ REACHABLE'
        google_results['Google Maps API']['note'] = 'Google Maps service is accessible'
    else:
        google_results['Google Maps API']['status'] = '⚠️ RESTRICTED'
        google_results['Google Maps API']['note'] = f'Status code: {response.status_code}'
except requests.exceptions.Timeout:
    google_results['Google Maps API']['status'] = '❌ TIMEOUT'
    google_results['Google Maps API']['note'] = 'Connection timeout'
except Exception as e:
    google_results['Google Maps API']['status'] = '❌ FAILED'
    google_results['Google Maps API']['note'] = str(e)

# Test Google Cloud connectivity
print("Testing Google Cloud connectivity...")
try:
    response = requests.head('https://cloud.google.com', timeout=5)
    if response.status_code < 400:
        google_results['Google Cloud Services']['status'] = '✅ REACHABLE'
        google_results['Google Cloud Services']['note'] = 'Google Cloud is accessible'
    else:
        google_results['Google Cloud Services']['status'] = '⚠️ RESTRICTED'
        google_results['Google Cloud Services']['note'] = f'Status code: {response.status_code}'
except requests.exceptions.Timeout:
    google_results['Google Cloud Services']['status'] = '❌ TIMEOUT'
    google_results['Google Cloud Services']['note'] = 'Connection timeout'
except Exception as e:
    google_results['Google Cloud Services']['status'] = '❌ FAILED'
    google_results['Google Cloud Services']['note'] = str(e)

# Test Firebase connectivity
print("Testing Firebase connectivity...")
try:
    response = requests.head('https://firebase.google.com', timeout=5)
    if response.status_code < 400:
        google_results['Firebase']['status'] = '✅ REACHABLE'
        google_results['Firebase']['note'] = 'Firebase is accessible'
    else:
        google_results['Firebase']['status'] = '⚠️ RESTRICTED'
        google_results['Firebase']['note'] = f'Status code: {response.status_code}'
except requests.exceptions.Timeout:
    google_results['Firebase']['status'] = '❌ TIMEOUT'
    google_results['Firebase']['note'] = 'Connection timeout'
except Exception as e:
    google_results['Firebase']['status'] = '❌ FAILED'
    google_results['Firebase']['note'] = str(e)

# Print Google results
for service, result in google_results.items():
    status = result['status']
    note = result.get('note', '')
    print(f"{service}: {status}")
    if note:
        print(f"  └─ {note}")

print()

# ============================================
# 3. DJANGO API ENDPOINTS TEST
# ============================================
print("🔌 DJANGO API ENDPOINTS TEST")
print("-" * 80)

api_endpoints = [
    '/api/auth/register/',
    '/api/auth/token/',
    '/api/users/',
    '/api/farms/',
    '/api/disease-detection/',
    '/api/weather/',
    '/api/alerts/',
]

print("Available API endpoints configured:")
for endpoint in api_endpoints:
    print(f"  ✓ {endpoint}")

print()

# ============================================
# 4. BACKEND SERVER STATUS
# ============================================
print("🖥️  BACKEND SERVER STATUS")
print("-" * 80)

try:
    response = requests.get('http://localhost:8000/api/', timeout=3)
    server_status = "✅ RUNNING"
    server_note = f"HTTP {response.status_code}"
except requests.exceptions.ConnectionError:
    server_status = "❌ NOT RUNNING"
    server_note = "Cannot connect to http://localhost:8000"
except requests.exceptions.Timeout:
    server_status = "❌ TIMEOUT"
    server_note = "Connection timeout"
except Exception as e:
    server_status = "⚠️ ERROR"
    server_note = str(e)

print(f"Django Development Server: {server_status}")
print(f"  └─ {server_note}")
print()

# ============================================
# 5. ENVIRONMENT CONFIGURATION
# ============================================
print("⚙️  ENVIRONMENT CONFIGURATION")
print("-" * 80)

from django.conf import settings

config_details = {
    'DEBUG Mode': settings.DEBUG,
    'Allowed Hosts': settings.ALLOWED_HOSTS,
    'Database Type': settings.DATABASES['default']['ENGINE'],
    'JWT Authentication': 'Enabled' if settings.REST_FRAMEWORK else 'Disabled',
    'CORS Enabled': bool(settings.CORS_ALLOWED_ORIGINS),
    'Email Backend': settings.EMAIL_BACKEND,
}

for key, value in config_details.items():
    if isinstance(value, bool):
        value = "✅ Yes" if value else "❌ No"
    elif isinstance(value, list):
        value = ", ".join(value) if value else "Not configured"
    print(f"{key}: {value}")

print()

# ============================================
# SUMMARY
# ============================================
print("=" * 80)
print("SUMMARY")
print("=" * 80)

summary = {
    'Database Connection': db_results['status'],
    'Google Services': 'Partially Configured' if any('✅' in str(v['status']) for v in google_results.values()) else 'Not Configured',
    'Django Backend': 'Ready' if db_results['status'].startswith('✅') else 'Check Required',
}

print()
for check, result in summary.items():
    print(f"✓ {check}: {result}")

print()
print("=" * 80)

# Additional Recommendations
print("📋 RECOMMENDATIONS")
print("-" * 80)
print("1. DATABASE: Configure connection pooling for production")
print("2. GOOGLE: Set up API keys for Maps, Cloud, OAuth if needed")
print("3. SECURITY: Enable HTTPS and secure settings in production")
print("4. MONITORING: Set up logging for database and API requests")
print("5. BACKUP: Configure database backups")
print()
